otaclient/app/ota_metadata.py

```diff
@@ -239,8 +186,155 @@ def _verify_certificate(self, certificate: bytes):
+    rootfs_directory: MetaFieldDescriptor[str] = MetaFieldDescriptor(str)
+    compressed_rootfs_directory: MetaFieldDescriptor[str] = MetaFieldDescriptor(str)
+    # metadata files definition
+    directory: MetaFieldDescriptor[MetaFile] = MetaFieldDescriptor(MetaFile)
+    symboliclink: MetaFieldDescriptor[MetaFile] = MetaFieldDescriptor(MetaFile)
+    regular: MetaFieldDescriptor[MetaFile] = MetaFieldDescriptor(MetaFile)
+    persistent: MetaFieldDescriptor[MetaFile] = MetaFieldDescriptor(MetaFile)
+    certificate: MetaFieldDescriptor[MetaFile] = MetaFieldDescriptor(MetaFile)
+
+    @classmethod
+    def parse_payload(cls, _input: Union[str, bytes]) -> "OTAMetadata":
+        # NOTE: in version1, payload is a list of dict
+        payload: List[Dict[str, Any]] = json.loads(_input)
+        if not payload or not isinstance(payload, list):
+            raise ValueError(f"invalid payload: {_input}")
+
+        res = cls()
+        for entry in payload:
+            # if entry is version, check version compatibility
+            if (
+                cls.VERSION_KEY in entry
+                and entry[cls.VERSION_KEY] != cls.SCHEME_VERSION
+            ):
+                logger.warning(f"metadata version is {entry['version']}.")
+            # parse other fields
+            for f in fields(cls):
+                if (fn := f.name) in entry:
+                    setattr(res, fn, entry)
+        return res
+
+    def get_metafiles(self) -> Iterator[MetaFile]:
+        for f in fields(self):
+            if (
+                (fd := getattr(self.__class__, f.name))
+                and isinstance(fd, MetaFieldDescriptor)
+                and fd.field_type is MetaFile
+            ):
+                yield getattr(self, f.name)
+
+    def get_image_data_url(self, base_url: str) -> str:
+        if getattr(self, "_data_url", None) is None:
+            self._data_url = urljoin_ensure_base(
+                base_url, f"{self.rootfs_directory.strip('/')}/"
+            )
+        return self._data_url
+
+    def get_image_compressed_data_url(self, base_url: str) -> str:
+        if getattr(self, "_compressed_data_url", None) is None:
+            self._compressed_data_url = urljoin_ensure_base(
+                base_url, f"{self.compressed_rootfs_directory.strip('/')}/"
+            )
+        return self._compressed_data_url
+
+    def get_download_url(
+        self, reg_inf: RegularInf, *, base_url: str
+    ) -> Tuple[str, Optional[str]]:
+        """
+        NOTE: compressed file is located under another OTA image remote folder
+
+        Returns:
+            A tuple of download url and zstd_compression enable flag.
+        """
+        # v2 OTA image, with compression enabled
+        # example: http://example.com/base_url/data.zstd/a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3.<compression_alg>
+        if (
+            compress_alg := reg_inf.compressed_alg
+        ) and compress_alg in cfg.SUPPORTED_COMPRESS_ALG:
```

Comment:
`compress_alg` is used in the if statement and within the if cluster, so I use `:=` here to avoid extra `.` operation.
But the if statement can be optimized as follow though:
```
    if (compress_alg := reg_inf.compressed_alg) in cfg.SUPPORTED_COMPRESS_ALG:
        ...
```

Reply:
What I am saying is assignment `compress_alg := reg_inf.compressed_alg` is redundant.

Reply:
https://github.com/tier4/ota-client/pull/172/files/8f60c3fc52fed4918d983a1006f053c5f15cd362#diff-4ee5c83d854eb84359277b1b90adc21f7a111fd258c4bf65a847307c1f327288R324
is using `reg_inf.compressed_alg`.

Reply:
Ah! I got it.

Reply:
Updated in [c72f8f0](https://github.com/tier4/ota-client/pull/172/commits/c72f8f083fb38227550bf3934b423565a2bd5332).